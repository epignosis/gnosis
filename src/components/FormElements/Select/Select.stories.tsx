import React from "react";
import { Story } from "@storybook/react";
import SelectComponent, { SelectProps } from "./Select";

export default {
  title: "components/Form Elements/Select",
  component: SelectComponent,
  args: {
    size: "md",
    label: "Choose a programming language",
    inline: false,
    id: "selectStory",
    className: "selectStory",
  },
  argTypes: {
    size: {
      control: {
        type: "select",
        options: ["md", "lg"],
      },
    },
    onChange: { action: "value changed" },
  },
  decorators: [
    (Story: Story): JSX.Element => (
      <div style={{ maxWidth: 400 }}>
        <Story />
      </div>
    ),
  ],
};

export const Select: Story<SelectProps> = (args) => (
  <SelectComponent id="programming-language" {...args}>
    <option value="rs">Rust</option>
    <option value="js">JavaScript</option>
    <option value="ts">TypeScript</option>
    <option value="go">GoLang</option>
    <option value="python">Python</option>
    <option value="php" disabled>
      PHP
    </option>
  </SelectComponent>
);
