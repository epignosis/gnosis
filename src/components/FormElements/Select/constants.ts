export const MAX_MENU_HEIGHT = 300;
export const OUTER_PLACEHOLDER = "Select...";
export const INNER_PLACEHOLDER = "Search...";
export const MIN_WIDTH = "5rem";
export const MAX_WIDTH = "25rem";

export const selectOptionsWithLevels = [
  {
    label: "category 2",
    value: "6",
    level: 0,
  },
  {
    label: "category1",
    value: "138",
    level: 1,
  },
  {
    label: "category 1vvv",
    value: "141",
    level: 2,
  },
  {
    label: "category 3xxx",
    value: "13123",
    level: 3,
  },
  {
    label: "category 1xxx",
    value: "140",
    level: 2,
  },
  {
    label: "category3",
    value: "139",
    level: 1,
  },
  {
    label: "My category",
    value: "32",
    level: 2,
  },
  {
    label: "Grand parent",
    value: "142",
    level: 0,
  },
  {
    label: "Parent",
    value: "143",
    level: 1,
  },
  {
    label: "Child Dos",
    value: "145",
    level: 2,
  },
  {
    label: "Child uno",
    value: "144",
    level: 2,
  },
];
