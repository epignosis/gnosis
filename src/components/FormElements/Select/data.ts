import { CustomOption } from "./types";

export interface ColourOption {
  value: string;
  label: string;
  color: string;
  isFixed?: boolean;
  isDisabled?: boolean;
  testId?: string;
}

export const colourOptions: ColourOption[] = [
  { value: "ocean", label: "Ocean", color: "#00B8D9", testId: "ocean" },
  { value: "blue", label: "Blue", color: "#0052CC", testId: "blue" },
  { value: "purple", label: "Purple", color: "#5243AA", testId: "purple" },
  { value: "red", label: "Red", color: "#FF5630", testId: "red" },
  { value: "orange", label: "Orange", color: "#FF8B00", testId: "orange" },
  { value: "yellow", label: "Yellow", color: "#FFC400", testId: "yellow" },
  { value: "green", label: "Green", color: "#36B37E", testId: "green" },
  { value: "forest", label: "Forest", color: "#00875A", testId: "forest" },
  { value: "slate", label: "Slate", color: "#253858", testId: "slate" },
  { value: "silver", label: "Silver", color: "#666666", testId: "silver" },
];

export interface FlavourOption {
  value: string;
  label: string;
  testId?: string;
}

export const flavourOptions: FlavourOption[] = [
  { value: "vanilla", label: "Vanilla", testId: "vanilla" },
  { value: "chocolate", label: "Chocolate", testId: "chocolate" },
  { value: "strawberry", label: "Strawberry", testId: "strawberry" },
  { value: "salted-caramel", label: "Salted Caramel", testId: "salted-caramel" },
];

export interface GroupedOption {
  label: string;
  options: ColourOption[] | FlavourOption[];
}

export const groupedOptions: GroupedOption[] = [
  {
    label: "Colours",
    options: colourOptions,
  },
  {
    label: "Flavours",
    options: flavourOptions,
  },
];

export const defaultOptions: CustomOption[] = [
  { label: "Rust", value: "rs", testId: "rust" },
  { label: "JavaScript is the best in the world", value: "js", testId: "javascript" },
  { label: "TypeScript", value: "ts", testId: "typescript" },
  { label: "GoLang", value: "go", testId: "golang" },
  { label: "Python", value: "python", testId: "python" },
  { label: "PHP", value: "php", testId: "php" },
  { label: "C++", value: "c++", testId: "c++" },
  { label: "C#", value: "c#", testId: "c#" },
  { label: "Java", value: "java", testId: "java" },
  { label: "Krust", value: "Krust", testId: "krust" },
  { label: "DontGo", value: "DontGo", hint: "Contact us", testId: "dontgo" },
  { label: "Cobol", value: "Cobol", testId: "cobol" },
  { label: "Haskell", value: "Haskell", testId: "haskell" },
  { label: "Erlang", value: "Erlang", testId: "erlang" },
  { label: "Perl", value: "Perl", testId: "perl" },
  { label: "Ruby", value: "Ruby", testId: "ruby" },
  { label: "Scala", value: "Scala", testId: "scala" },
  { label: "Kotlin", value: "Kotlin", testId: "kotlin" },
  { label: "Disabled option", value: "disabled", testId: "disabled-option", disabled: true },
];

export const menuMaxWidthOptions: CustomOption[] = [
  {
    label:
      "Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book",
    value: "lorem",
    testId: "lorem",
  },
  { label: "JavaScript is the best in the world", value: "js", testId: "javascript" },
  { label: "TypeScript", value: "ts", testId: "typescript" },
  { label: "GoLang", value: "go", testId: "golang" },
  { label: "Python", value: "python", testId: "python" },
  { label: "PHP", value: "php", testId: "php" },
  { label: "C++", value: "c++", testId: "c++" },
  { label: "C#", value: "c#", testId: "c#" },
  { label: "Java", value: "java", testId: "java" },
  {
    label:
      "Disabled Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book",
    value: "lorem disabled",
    testId: "lorem-disabled",
    disabled: true,
  },
];
