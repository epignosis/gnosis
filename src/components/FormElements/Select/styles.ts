import { css, SerializedStyles, Theme } from "@emotion/react";
import { inputBaseStyles, inputContainerBaseStyles } from "../styles";
import { InputSize } from "@components/FormElements/Input/Input";

export const selectContainer = (
  { formElements }: Theme,
  { size }: { size: InputSize },
): SerializedStyles => css`
  ${inputContainerBaseStyles({ block: true })}

  label {
    margin: 0 0 0.5rem 0.5rem;
  }

  select {
    ${inputBaseStyles(formElements, { block: true, size })};
    background-color: transparent;
    padding-right: 2.5rem;
    appearance: none;

    &:hover,
    &:focus {
      border: 1px solid ${formElements.input.borderHoverColor};
    }
  }

  .select-input-wrapper {
    position: relative;
    width: 100%;
    background-color: ${formElements.input.background};
    z-index: 1;

    .carret-wrapper {
      position: absolute;
      right: 0.75rem;
      height: 100%;
      display: inline-flex;
      align-items: center;
      z-index: -1;

      svg {
        color: #000;
      }
    }
  }
`;
