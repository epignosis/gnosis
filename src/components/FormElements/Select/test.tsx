import React, { FC, useRef } from "react";
import { SelectInstance } from "react-select";
import CustomSelect from "./Select";
import { CustomOptionType, CustomTypeParam } from "./types";

const Test: FC = () => {
  const handleOnChange = (v: CustomTypeParam) => {
    // eslint-disable-next-line no-console
    console.log("test", v);
  };

  const ref = useRef<SelectInstance<CustomOptionType>>(null);

  return (
    <CustomSelect
      ref={ref}
      id="sorting"
      containerAttrs={{ className: "select-wrapper" }}
      isRtl={false}
      onChange={handleOnChange}
      options={[
        {
          label: "name",
          value: "name",
        },
      ]}
    />
  );
};

export default Test;
