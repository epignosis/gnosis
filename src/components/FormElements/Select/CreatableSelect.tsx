import React, { ForwardRefRenderFunction, forwardRef } from "react";
import { SelectInstance } from "react-select";
import CreatableSelect from "react-select/creatable";
import { SerializedStyles } from "@emotion/react";
import classNames from "classnames";
import Label from "../Label/Label";
import { resolveStyles, selectContainer } from "./styles";
import CustomMenuList from "./components/CustomMenuList";
import { CustomOption, CustomSelectProps } from "./types";

const MAX_MENU_HEIGHT = 300;
const MIN_WIDTH = "5rem";
const MAX_WIDTH = "25rem";

const containerClassNames = (status: string, size: string) =>
  classNames({
    [`control-${size}`]: true,
    valid: status === "valid",
    error: status === "error",
  });

const Select: ForwardRefRenderFunction<
  SelectInstance<CustomOption>,
  CustomSelectProps<CustomOption, boolean>
> = (props, forwardedRef) => {
  const {
    id = "",
    label,
    size = "md",
    status = "valid",
    hasInnerSearch = false,
    isMulti = false,
    maxMenuHeight = MAX_MENU_HEIGHT,
    minWidth = MIN_WIDTH,
    maxWidth = MAX_WIDTH,
    ...rest
  } = props;
  const hasLabel = Boolean(label);

  const styles = resolveStyles(size, hasInnerSearch);

  return (
    <div
      css={(theme): SerializedStyles =>
        selectContainer(theme, {
          size,
          minWidth,
          maxWidth,
          inline: false,
          isInlineFlex: false,
          hasInnerSearch: false,
        })
      }
    >
      {hasLabel && (
        <Label htmlFor={id} aria-labelledby={id}>
          {label}
        </Label>
      )}
      <div className="select-input-wrapper" data-testid="custom-react-select">
        <CreatableSelect
          {...rest}
          ref={forwardedRef}
          blurInputOnSelect={!isMulti}
          closeMenuOnSelect={!isMulti}
          isMulti={isMulti}
          classNames={{
            control: () => containerClassNames(status, size),
            option: () => `option-${size}`,
          }}
          components={{
            IndicatorSeparator: () => null,
            MenuList: CustomMenuList,
          }}
          maxMenuHeight={maxMenuHeight}
          styles={styles}
        />
      </div>
    </div>
  );
};

export default forwardRef(Select);
