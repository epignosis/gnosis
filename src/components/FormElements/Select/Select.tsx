import React, { forwardRef, ForwardRefRenderFunction } from "react";
import { SerializedStyles } from "@emotion/react";
import classNames from "classnames";
import Label from "../Label/Label";
import { selectContainer } from "./styles";
import { CaretDownSVG } from "@icons/core";
import { ExtendableProps } from "types/common";

export type SelectProps = ExtendableProps<
  React.SelectHTMLAttributes<HTMLSelectElement>,
  {
    status?: "valid" | "error";
    size?: "md" | "lg";
    label?: string;
    inline?: boolean;
    onChange?: (selectedValue: string) => void;
  }
>;

const Select: ForwardRefRenderFunction<HTMLSelectElement, SelectProps> = (props, forwardedRef) => {
  const {
    id,
    status = "valid",
    size = "md",
    label,
    inline = false,
    onChange,
    className = "",
    children,
    ...rest
  } = props;
  const hasLabel = Boolean(label);
  const containerClassNames = classNames({
    valid: status === "valid",
    error: status === "error",
    inline: hasLabel && inline,
    [className]: className,
  });

  return (
    <div
      css={(theme): SerializedStyles => selectContainer(theme, { size })}
      className={containerClassNames}
    >
      {hasLabel && <Label htmlFor={id}>{label}</Label>}
      <div className="select-input-wrapper">
        <select
          ref={forwardedRef}
          id={id}
          onChange={onChange && ((e): void => onChange(e.target.value))}
          {...rest}
        >
          {children}
        </select>
        <span className="carret-wrapper">
          <CaretDownSVG height={24} />
        </span>
      </div>
    </div>
  );
};

export default forwardRef(Select);
