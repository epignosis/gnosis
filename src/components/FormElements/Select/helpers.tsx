import React from "react";
import classNames from "classnames";
import AsyncSelect from "react-select/async";
import ReactSelect, { FormatOptionLabelContext } from "react-select";
import CreatableSelect from "react-select/creatable";
import { SerializedStyles } from "@emotion/react";
import { CarretArrowRight } from "../../../icons/";
import { SelectType, CustomSelectProps, CustomOption } from "./types";
import { customLabelStyles } from "./styles";

export const containerClassNames = (status: string, size: string) =>
  classNames({
    [`control-${size}`]: true,
    valid: status === "valid",
    error: status === "error",
  });

export const formatOptionLabel = (
  { label = "", level = 0 },
  { context }: { context: FormatOptionLabelContext },
): JSX.Element => {
  if (!level || context === "value") return <>{label}</>;

  const isRtl = document.dir === "rtl";
  const rotation = isRtl ? 180 : 0;

  return (
    <span css={(): SerializedStyles => customLabelStyles({ level })}>
      {level ? (
        <CarretArrowRight height={10} style={{ transform: `rotate(${rotation}deg)` }} />
      ) : null}
    </span>
  );
};

export const renderSelect = (
  type: SelectType,
  customSelectProps: CustomSelectProps<CustomOption, boolean>,
) => {
  const { loadOptions, ...rest } = customSelectProps;

  switch (type) {
    case "creatable":
      return (
        <CreatableSelect
          {...rest}
          createOptionPosition="first"
          formatOptionLabel={formatOptionLabel}
        />
      );
    case "async":
      return <AsyncSelect {...rest} loadOptions={loadOptions} />;
    default:
      return <ReactSelect {...rest} formatOptionLabel={formatOptionLabel} />;
  }
};
