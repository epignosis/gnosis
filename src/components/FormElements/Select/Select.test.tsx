import React from "react";
import userEvent from "@testing-library/user-event";
import faker from "faker";
import Select from "./Select";
import { screen, render } from "@test-utils/render";

const OPTIONS = [
  {
    label: faker.random.words(),
    value: faker.random.alphaNumeric(),
  },
  {
    label: faker.lorem.words(),
    value: faker.lorem.slug(),
  },
  {
    label: faker.git.commitMessage(),
    value: faker.git.shortSha(),
  },
];

describe("<Select />", () => {
  it("renders correctly", () => {
    const labelTxt = faker.commerce.department();
    render(
      <Select id={faker.random.alpha()} label={labelTxt}>
        {OPTIONS.map(({ label, value }) => (
          <option key={value} value={value}>
            {label}
          </option>
        ))}
      </Select>,
    );

    const select = screen.getByLabelText(labelTxt);
    const options = screen.getAllByRole("option");

    expect(select).toBeInTheDocument();
    expect(options).toHaveLength(3);
  });

  it("selects a value", () => {
    const labelTxt = faker.commerce.department();
    render(
      <Select id={faker.random.alpha()} label={labelTxt}>
        {OPTIONS.map(({ label, value }) => (
          <option key={value} value={value}>
            {label}
          </option>
        ))}
      </Select>,
    );

    const select = screen.getByLabelText(labelTxt);

    expect(select).toHaveValue(OPTIONS[0].value);

    userEvent.selectOptions(select, OPTIONS[1].value);

    expect(select).toHaveValue(OPTIONS[1].value);
  });

  it("matches snapshot", () => {
    const { container } = render(
      <Select id="test-select" label="Test select input">
        <option value="rs">Rust</option>
        <option value="js">JavaScript</option>
        <option value="ts">TypeScript</option>
      </Select>,
    );

    expect(container).toMatchSnapshot();
  });
});
