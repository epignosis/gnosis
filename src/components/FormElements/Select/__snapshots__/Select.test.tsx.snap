// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Select /> matches snapshot 1`] = `
<div>
  <div
    class="valid container-class css-n8tftf-inputContainerBaseStyles-inputContainerBaseStyles-inputBaseStyles-inputBaseStyles-selectContainer-selectContainer"
    id="container-id"
  >
    <label
      aria-labelledby="select-id"
      class="css-xs6o3g-label-label"
      for="select-id"
    >
      Test select input
    </label>
    <div
      class="select-input-wrapper"
    >
      <select
        class="normal-select"
        id="select-id"
      >
        <option
          value="rs"
        >
          Rust
        </option>
        <option
          value="js"
        >
          JavaScript
        </option>
        <option
          value="ts"
        >
          TypeScript
        </option>
      </select>
      <span
        class="carret-wrapper"
      >
        <svg-file-stub
          height="24"
        />
      </span>
    </div>
  </div>
</div>
`;
